mjpc/tasks/h1/walk/walk.cc

```diff
@@ -0,0 +1,238 @@
+  // mju_scl(com_vel, com_vel, 0.5, 2);
+
+  double com_vel[2];
+  mju_copy(com_vel,subcomvel,2); //subcomvel is the velocity of the robot's CoM
+
+  // face forward
+  residual[counter++] =
+      standing * (mju_dot(forward, forward_target, 2) - 1);
+
+  // walk forward
+  residual[counter++] =
+      standing * (mju_dot(com_vel, forward_target, 2) - parameters_[1]);
+
+  // ----- move feet ----- //
+  double* foot_right_vel = SensorByName(model, data, "foot_right_velocity");
+  double* foot_left_vel = SensorByName(model, data, "foot_left_velocity");
+  double move_feet[2];
+  mju_copy(move_feet, com_vel, 2);
+  mju_addToScl(move_feet, foot_right_vel, -0.5, 2);
+  mju_addToScl(move_feet, foot_left_vel, -0.5, 2);
+
+  mju_copy(&residual[counter], move_feet, 2);
+  mju_scl(&residual[counter], &residual[counter], standing, 2);
+  counter += 2;
+
+  // ----- control ----- //
+  mju_copy(&residual[counter], data->ctrl, model->nu);
+  counter += model->nu;
+
+  // ----- feet distance ----- //
+  double feet_axis[2];
+  mju_copy(feet_axis, foot_right, 2);
+  mju_addToScl(feet_axis, foot_left, -1, 2);
+  double feet_distance = mju_norm(feet_axis, 2);
+  residual[counter] = feet_distance - parameters_[2];
+  counter += 1;
+
+  // ----- feet cross ----- //
+  double *left_foot_left_axis = SensorByName(model, data, "foot_left_left");
+  mju_sub3(vec, foot_right, foot_left);
+  residual[counter++] = mju_dot3(vec, left_foot_left_axis) - 0.2;
+  
+
+  // ----- slippage ----- //
+  double* foot_right_ang_velocity = SensorByName(model, data, "foot_right_ang_velocity");
+  double* foot_left_ang_velocity = SensorByName(model, data, "foot_left_ang_velocity");
+  double* right_foot_xbody = SensorByName(model, data, "foot_right_xbody");
+  double* left_foot_xbody = SensorByName(model, data, "foot_left_xbody");
+  
+  
+  residual[counter++] = (tanh(-(right_foot_xbody[2]-0.0645)/0.001)+1)*0.5*foot_right_ang_velocity[2];
+  residual[counter++] = (tanh(-(left_foot_xbody[2]-0.0645)/0.001)+1)*0.5*foot_left_ang_velocity[2];
+
+  // sensor dim sanity check
+  // TODO: use this pattern everywhere and make this a utility function
+  int user_sensor_dim = 0;
+  for (int i = 0; i < model->nsensor; i++) {
+    if (model->sensor_type[i] == mjSENS_USER) {
+      user_sensor_dim += model->sensor_dim[i];
+    }
+  }
+  if (user_sensor_dim != counter) {
+    mju_error_i(
+        "mismatch between total user-sensor dimension"
+        "and actual length of residual %d",
+        counter);
+  }
+}
+
+double Walk::ResidualFn::StepHeight(double time, double footphase,
+                                             double duty_ratio) const {
+  double angle = fmod(time + mjPI - footphase, 2*mjPI) - mjPI;
+  double value = 0;
+  if (duty_ratio < 1) {
+    angle *= 0.5 / (1 - duty_ratio);
+    value = mju_cos(mju_clip(angle, -mjPI/2, mjPI/2));
+  }
+  return mju_abs(value) < 1e-6 ? 0.0 : value;
+}
+
+}  // namespace mjpc::h1
```

Comment:
Please format this file, see this [reference](https://github.com/google-deepmind/mujoco_mpc/blob/main/docs/CONTRIBUTING.md#code-style).